lua/entities/lvs_base/sh_weapons.lua

```diff
@@ -268,7 +269,12 @@ if SERVER then
 
 			local ShootDelay = (CurWeapon.Delay or 0)
 
-			self:SetNextAttack( T + ShootDelay )
+			-- When firing full auto we want to base our next attack time off of our last attack time and not the current time because that's rounded to ticks and thus inaccurate, causing mishaps on low tickrate.
+			if CurWeapon._NextFire and T - CurWeapon._NextFire < 1e-4 then
+				self:SetNextAttack( CurWeapon._NextFire + ShootDelay )
+			else
+				self:SetNextAttack( T + ShootDelay )
+			end
```

Comment:
At some point there are just limits in what low tick rate can do when it comes to firerate. Wouldnt this change just make weapons fire faster and faster until they just fire on tick if CurTime() progresses faster than fire delay added here? This needs to be checked

Reply:
btw: gunner has a weapon system too:
https://github.com/SpaxscE/lvs_base/blob/main/lua/entities/lvs_base_gunner/init.lua#L229

Reply:
i believe a better method to fix this would be to modify ShootDelay itself. Once it reaches the critical value just make it fire on tick instead.

Reply:
> Wouldnt this change just make weapons fire faster and faster until they just fire on tick if CurTime() progresses faster than fire delay added here? This needs to be checked
> 
> Fast firing weapons are expected to break if their shoot delay is getting closer to tick interval. At some point there are just limits in what low tick rate can do when it comes to firerate

It does technically make the weapons fire 'faster', but that's only because with your original code, the fire rate was actually slower than intended. [This video](https://youtu.be/7puuYqq_rgw) nicely explains the issue of setting the next attack time relative to CurTime instead of itself when firing full auto weapons.

As for weapons breaking as their fire delay approaches tick interval, this issue can be fixed by simply replacing the if statement with a while loop or running the attack code again, if we shoot a bullet and our next attack time is still within the window of the current tick, we can simply fire multiple bullets, in a single frame, Counter-Strike weapons and most game developers in general already do this in most of their games as can be seen [here](https://github.com/perilouswithadollarsign/cstrike15_src/blob/f82112a2388b841d72cb62ca48ab1846dfcc11c8/game/shared/cstrike15/weapon_csbase.cpp#L1508) and is documented [here](https://developer.valvesoftware.com/wiki/FireBulletsInfo_t).